File: ruby/lib/use_case/order_approval_use_case.rb
# frozen_string_literal: true

require_relative '../domain/order_status'

class OrderApprovalUseCase
  class ShippedOrdersCannotBeChangedError < StandardError; end
  class RejectedOrderCannotBeApprovedError < StandardError; end
  class ApprovedOrderCannotBeRejectedError < StandardError; end

  def initialize(order_repository)
    @order_repository = order_repository
  end

  # @param request [OrderApprovalRequest]
  def run(request)
    order = @order_repository.get_by_id(request.order_id)

    raise ShippedOrdersCannotBeChangedError if order.status == OrderStatus::SHIPPED

    raise RejectedOrderCannotBeApprovedError if request.approved && order.status == OrderStatus::REJECTED

    raise ApprovedOrderCannotBeRejectedError if !request.approved && order.status == OrderStatus::APPROVED

    order.status = request.approved ? OrderStatus::APPROVED : OrderStatus::REJECTED

    @order_repository.save(order)
  end
end

File: python/src/use_case/order_approval_use_case.py
from ..domain.order_status import OrderStatus
from ..repository.order_repository import OrderRepository
from .exceptions import ShippedOrdersCannotBeChangedException, \
    RejectedOrderCannotBeApprovedException, ApprovedOrderCannotBeRejectedException
from .order_approval_request import OrderApprovalRequest


class OrderApprovalUseCase:
    def __init__(self, order_repository: OrderRepository) -> None:
        self._order_repository = order_repository

    def run(self, request: OrderApprovalRequest) -> None:
        order = self._order_repository.get_by_id(request.order_id)

        if order.status == OrderStatus.SHIPPED:
            raise ShippedOrdersCannotBeChangedException()

        if request.approved and order.status == OrderStatus.REJECTED:
            raise RejectedOrderCannotBeApprovedException()

        if not request.approved and order.status == OrderStatus.APPROVED:
            raise ApprovedOrderCannotBeRejectedException()

        order.status = OrderStatus.APPROVED if request.approved else OrderStatus.REJECTED
        self._order_repository.save(order)
